### PR TITLE
Allow selection to produce columns, closes #44.

### LINKED ISSUE #44: Allow production of columns with selection through actual producers
Right now, the way that new columns can be created is somewhat two-fold. On the one hand, `SelectionResult`'s can contain new columns that will be injected into saved selection masks. On the other hand, we have the generic `producer` interface that does the same but a whole lot nicer. At the same time, `selector`'s have a `produces` member which is not utilized whatsover.

We should allow producers being called from within selectors and allow selection functions to return either

- a single value, being the `SelectionResult` (but a variant **without** new columns), or
- a 2-tuple, being the `SelectionResult` and additional columns in a separate array.

This would make much more sense.

@pkausw @dsavoiu

### PR DESCRIPTION
**Pull request by riga**
This PR tackles the ambiguity described in #44.

>  Right now, the way that new columns can be created is somewhat two-fold. On the one hand, SelectionResult's can contain new columns that will be injected into saved selection masks. On the other hand, we have the generic producer interface that does the same but a whole lot nicer. At the same time, selector's have a produces member which is not utilized whatsover.

Changes in this PR:

- `SelectionResult` does not longer have a field for new columns.
- Selectors now can return a single `SelectionResult` or a 2-tuple (`SelectionResult`, events)
- When the selector configured for the `SelectEvents` task produces columns, `SelectEvents` will have a third output representing the produced columns (besides the selection masks and the stats file). **However**, we should only use that feature if it's really necessary (e.g. for saving columns that are required for the selection step, but are otherwise expensive to reproduce) and adhere to the `ProduceColumns` task if possible (see the task structure in #25).
- Current selectors are adapted to the above structure.

At some point, we should make some of the selectors which are actually producing sth, actual producers :)

@mafrahm

### CODE REVIEW

File: ap/tasks/reduction.py [hunk 1]
```diff
@@ -76,30 +76,38 @@ def run(self):
             nano_file = input_file.load(formatter="uproot")
 
         # iterate over chunks of events and diffs
+        input_paths = [nano_file]
+        input_paths.append(inputs["sel"]["res"].path)
+        input_paths.extend([inp.path for inp in inputs["calib"]])
+        if self.selector_func.produced_columns:
+            input_paths.append(inputs["sel"]["cols"].path)
```
Comment (pkausw) [inline ap/tasks/reduction.py:83]: See other comments regarding descriptive power of keyword arguments

```suggestion
            input_paths.append(inputs["selection"]["columns"].path)
```

File: ap/tasks/reduction.py [hunk 2]
```diff
@@ -76,30 +76,38 @@ def run(self):
...
+                events = events[event_mask]
 
                 # apply all object masks whose names are present
+                # TODO: this should not be done automagically based on names as there might be
+                #       multiple collections of the same object type
                 for name in sel.objects.fields:
                     if name in events.fields:
                         # apply the event mask to the object mask first
-                        object_mask = sel.objects[name][sel.event]
+                        object_mask = sel.objects[name][event_mask]
```
Comment (pkausw) [inline ap/tasks/reduction.py:110]: This depends on wether the selection masks in `selection` can be stacked, right? But I guess this shouldn't break even if the selections were stacked before 🤔

File: ap/tasks/selection.py [hunk 3]
```diff
@@ -46,25 +46,32 @@ def requires(self):
         return reqs
 
     def output(self):
-        return {
+        outputs = {
             "res": self.local_target(f"results_{self.branch}.parquet"),
             "stat": self.local_target(f"stats_{self.branch}.json"),
         }
```
Comment (pkausw) [inline ap/tasks/selection.py:52]: see comment in file `ap/tasks/reduction.py` regarding descriptive power of keys of dictionary. Wouldn't it be better to make it clearer?
```suggestion
        outputs = {
            "result": self.local_target(f"results_{self.branch}.parquet"),
            "stats": self.local_target(f"stats_{self.branch}.json"),
        }

File: ap/tasks/selection.py [hunk 4]
```diff
@@ -46,25 +46,32 @@ def requires(self):
...
 
+        # add additional columns in case the selector produces some
+        if self.selector_func.produced_columns:
+            outputs["cols"] = self.local_target(f"columns_{self.branch}.parquet")
```
Comment (pkausw) [inline ap/tasks/selection.py:56]: See 'descriptive power of keys' above
```suggestion
            outputs["columns"] = self.local_target(f"columns_{self.branch}.parquet")
```

**Review by pkausw — CHANGES_REQUESTED** [review]

File: ap/tasks/reduction.py [hunk 5]
```diff
@@ -76,30 +76,38 @@ def run(self):
             nano_file = input_file.load(formatter="uproot")
 
         # iterate over chunks of events and diffs
+        input_paths = [nano_file]
+        input_paths.append(inputs["sel"]["res"].path)
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/tasks/reduction.py]: Yep, added that ✓

File: ap/tasks/selection.py (hunk 3 above)
Comment (riga) [reply ap/tasks/selection.py]: Yep, added that ✓

File: ap/tasks/selection.py (hunk 4 above)
Comment (riga) [reply ap/tasks/selection.py]: Yep, added that ✓

File: ap/tasks/reduction.py (hunk 2 above)
Comment (pkausw) [reply ap/tasks/reduction.py:110]: OK :+1:

**Pull request merged**